core/src/main/java/com/linecorp/armeria/client/Endpoint.java

```diff
@@ -578,61 +576,49 @@ public <T> T attr(AttributeKey<T> key) {
      */
     public <T> Endpoint withAttr(AttributeKey<T> key, @Nullable T value) {
         requireNonNull(key, "key");
-        if (value == null && attributes == null) {
-            return this;
+        if (attributes == null) {
+            if (value == null) {
+                return this;
+            }
+            return withAttrs(Attributes.of(key, value));
         }
-        if (attributes != null && attributes.attr(key) == value) {
+
+        if (attributes.attr(key) == value) {
             return this;
+        } else {
+            final AttributesBuilder attributesBuilder = attributes.toBuilder();
+            if (value == null) {
+                attributesBuilder.remove(key);
```

Comment:
Q) May I ask the rationale behind using `remove` instead of `set(key, null)`? (since they behave slightly differently depending on whether the parent has the key)

Reply:
Originally, I designed `set(key, value)` not to allow null. I refactored the API design and now it looks better to use `set(key, null)` instead of `remove(key)`.